### ISSUE TITLE
Add post-training validation for root embeddings

### ISSUE THREAD

**Issue by marctjones** [marctjones/klareco#424]
## Problem

Current training script saves model but provides minimal validation:
- Vocabulary size
- Embedding dimension  
- Best correlation (from training)
- Training epochs

**Missing**:
- Semantic similarity validation (do similar roots have similar embeddings?)
- Coverage check (are all clean_roots.json roots learned?)
- Fundamento root quality check
- Anti-collapse metrics (are embeddings collapsed to single point?)

**Impact**: Can't detect training failures without manually inspecting embeddings.

## Proposed Enhancement

Add post-training validation script that runs automatically:

```python
# scripts/validate_root_embeddings.py

def validate_root_embeddings(model_path, clean_vocab_path):
    """
    Validate trained root embeddings for quality and correctness.
    
    Checks:
    1. Vocabulary coverage (all clean roots present?)
    2. Semantic similarity (mal- pairs, synonyms, antonyms)
    3. Embedding diversity (no collapse?)
    4. Fundamento coverage (all Fundamento roots learned?)
    """
    
    # Load model
    checkpoint = torch.load(model_path)
    embeddings = checkpoint['embeddings']
    vocab = checkpoint['vocabulary']
    
    # Check 1: Vocabulary coverage
    with open(clean_vocab_path) as f:
        clean_vocab = json.load(f)
    
    missing = set(clean_vocab.keys()) - set(vocab.keys())
    coverage = 100 * (1 - len(missing) / len(clean_vocab))
    
    print(f"Vocabulary Coverage: {coverage:.1f}%")
    if missing:
        print(f"  Missing roots: {len(missing)}")
        print(f"  Examples: {list(missing)[:10]}")
    
    # Check 2: Semantic similarity validation
    # Test mal- opposites (mal-bona vs bona should be dissimilar)
    # Test synonyms from ReVo (should be similar)
    
    # Check 3: Embedding diversity (anti-collapse)
    mean_similarity = compute_mean_pairwise_similarity(embeddings)
    print(f"Mean pairwise similarity: {mean_similarity:.4f}")
    if mean_similarity > 0.5:
        print("  ⚠️  WARNING: Possible embedding collapse!")
    
    # Check 4: Fundamento coverage
    fundamento_missing = check_fundamento_coverage(vocab)
    print(f"Fundamento coverage: {100 * (1 - len(fundamento_missing)/FUNDAMENTO_COUNT):.1f}%")
    
    return {
        'coverage': coverage,
        'mean_similarity': mean_similarity,
        'fundamento_coverage': fundamento_coverage,
        'missing_roots': list(missing),
    }
```

## Integration into train_roots.sh

Add validation step after training completes:

```bash
if [ -f "$OUTPUT_DIR/best_model.pt" ]; then
    echo -e "${GREEN}Model saved to: $OUTPUT_DIR/best_model.pt${NC}"
    
    # NEW: Run validation
    echo ""
    echo -e "${YELLOW}=== Running Post-Training Validation ===${NC}"
    python scripts/validate_root_embeddings.py \
        --model "$OUTPUT_DIR/best_model.pt" \
        --clean-vocab "$CLEAN_VOCAB" \
        --fundamento-roots "$FUNDAMENTO_ROOTS"
    
    # Show existing model info...
fi
```

## Acceptance Criteria

- [ ] Validation script checks vocabulary coverage
- [ ] Semantic similarity tests (mal- pairs, synonyms)
- [ ] Anti-collapse detection (mean similarity < 0.5)
- [ ] Fundamento coverage check (100% expected)
- [ ] Validation runs automatically after training
- [ ] Results saved to `validation_results.json`
- [ ] Clear warnings if quality issues detected

## Expected Output

```
=== Post-Training Validation ===

Vocabulary Coverage: 98.7%
  Missing roots: 167
  Examples: ['ekzotik', 'algoritm', 'pixel', ...]

Semantic Similarity Tests:
  mal- opposites: 23/25 passed ✓
  ReVo synonyms: 187/200 similar ✓
  ReVo antonyms: 95/100 dissimilar ✓

Embedding Diversity:
  Mean pairwise similarity: 0.23 ✓
  No collapse detected

Fundamento Coverage: 100.0% ✓
  All 2,642 Fundamento roots present

Overall: PASSED ✓
```

## Files to Create

- `scripts/validate_root_embeddings.py` (new validation script)

## Files to Update

- `scripts/train_roots.sh` (add validation step after training)

## Related

- Anti-collapse metrics already tracked during affix training
- ReVo semantic relations can provide ground truth for similarity tests

**Comment by marctjones** [comment]
### Validation Results (2026-01-12)

Ran validation script on trained model:

**Overall: PRODUCTION-READY ✓**
- Semantic Similarity: 7/7 PASS
- Cluster Coherence: 5/5 PASS  
- Fundamento Coverage: 56/56 PASS (100%)
- No Embedding Collapse: PASS (mean_sim=0.016)
- Anti-Patterns: 2/4 PASS (minor: bel↔danĝer=0.472)

**Key Metrics:**
- Correlation: 0.8017
- Accuracy: 0.9027
- Vocab: 10,819 roots

**Strong Performance:**
- pens↔sci (think/know): 0.758
- Family cluster: 0.607 avg similarity
- Communication cluster: 0.619 avg similarity

See note #427 for full validation report.
